
<a href="https://ibb.co/5WJ92c2C"><img src="https://i.ibb.co/HD92XgXR/Screenshot-2026-07-24-at-14-50-56.png" alt="Screenshot-2026-07-24-at-14-50-56" border="0"></a>

# The AI Lab Partner: Agents for Biological Discovery

**In this tutorial, we are going to tackle a real-world problem together: engineering a protein to be more thermostable**. If successful in the real world, this kind of engineering could eliminate the need for fragile cold-chain storage in regions like West Africa!

By the end of our time together, you will have built an automated "AI Scientist" that can iteratively design, test, and refine protein mutations on its own.

### Prerequisites

Just basic Python. Absolutely no biology background is required!

### 💡 A Note on the Code

> The Python code in this notebook is **not** the main focus of the lesson. Instead, our primary goal is to teach the **agentic workflow**—the art of prompting an AI agent to generate, run, and evaluate complex code on your behalf. If you don't fully understand every single line of code, that is perfectly okay! Focus on learning how to guide the AI's reasoning. 😊

### 🤖 Colab and Gemini or Antigravity

In this tutorial, you can use either the Gemini app or Gemini within Antigravity to generate code snippets, which you will then run in Colab.

This workflow may seem slightly indirect, particularly because Antigravity can execute code itself, so Colab is not always strictly necessary. However, we use Colab for two main reasons:

- Antigravity’s free tier can reach its execution quota quite quickly.
- We need a reliable set of backup cells that we know will work. Each student’s agent may generate slightly different code, so results can vary.

You’re encouraged to run the code generated by your agent. However, if it does not work, you’ll need to debug it with the agent—or use the provided backup cells. 🙂

### Tutorial contents

As a reminder, this tutorial's structure is:

<a href="https://ibb.co/QvTLkrN3"><img src="https://i.ibb.co/yFZtpWyc/Screenshot-2026-07-24-at-14-52-19.png" alt="Screenshot 2026 07 24 at 14 52 19" border="0"></a>

**Tutorial authors**: This session was co-written by Natasha Latysheva, Adam Kosiorek, Kyle Taylor, Sebastian Bodenstein, and Alexander Karollus.




## Phase 1: Discovery, Data Fetching, and Visualisation

*Focus: Topic exploration, automated data retrieval via APIs, and initial protein structure visualisation.*


### 🔍 1.1 Discovery Prompt

**Prompt:**
> I am a researcher at the Pan-Atlantic University in Lagos interested in protein stability and predicting it from sequence. Can you recommend a specific target protein relevant to health challenges in my country, where improving thermal stability would significantly benefit regions with fragile cold-chain infrastructure? Ideally, it would be something which could be worked on without specific safety precautions or a BSL2+ laboratory. Additionally, it should be less than 300 amino acids long, to make computational analysis (e.g. folding etc) as simple as possible.


<details>
<summary>🤖 Show Example Output</summary>

Here are four excellent candidate proteins that directly address critical public health challenges in Nigeria, are highly sensitive to cold-chain infrastructure failures, are under 300 amino acids long, and can be safely studied computationally or expressed in standard Biosafety Level 1 (BSL-1) laboratory environments.

1. **Plasmodium falciparum Merozoite Surface Protein 1 (PfMSP1-19)**
   - **Health Challenge:** Malaria is a leading cause of morbidity and mortality in Nigeria, with over 60 million cases annually. PfMSP1-19 is a primary antigen candidate for blood-stage malaria vaccines.
   - **Cold-Chain Impact:** Developing countries require thermostable vaccines that can be delivered to remote rural clinics without active refrigeration. Additionally, expressing PfMSP1-19 in simple mesophilic hosts (like E. coli or yeast) often leads to misfolding due to its complex disulfide network. Engineering it for increased structural stability and robustness could greatly improve vaccine storage and yield.
   - **Size:** 94 amino acids (very small and computationally highly tractable).
   - **Structure:** It consists of two epidermal growth factor (EGF)-like domains stabilized by 6 disulfide bonds.
   - **PDB Reference:** 1OB1 (PfMSP1-19 in complex with an antibody Fab fragment) or 6ZBC.
   - **Safety:** Recombinant subunit expression and computational design are entirely BSL-1.

2. **Single-Chain Insulin (SCI)**
   - **Health Challenge:** Non-communicable diseases like diabetes are rising rapidly in urban and rural Nigeria.
   - **Cold-Chain Impact:** Standard insulin is highly susceptible to heat-induced fibrillation (aggregation) and chemical degradation. For patients in hot climates who lack reliable electricity and refrigeration, keeping insulin cold is a constant challenge. There is a major research focus on designing single-chain insulin (SCI) analogs that substitute the native two-chain structure with a short linker, rendering the monomer structurally resilient to high temperatures (up to 45°C) without losing receptor binding activity.
   - **Size:** ~57 to 60 amino acids (A-chain and B-chain joined by a 6–7 amino acid linker).
   - **Structure:** Small, helical structure stabilized by 3 disulfide bonds.
   - **PDB Reference:** 2JZQ (an engineered monomeric single-chain insulin) or 5CO7.
   - **Safety:** insulin is a non-infectious, therapeutic hormone; both modeling and wet-lab work can be conducted in BSL-1.

3. **Neisseria meningitidis Factor H Binding Protein (fHbp)**
   - **Health Challenge:** Nigeria lies within the African Meningitis Belt, experiencing seasonal outbreaks of bacterial meningitis. Vaccines targeting fHbp are vital for protecting vulnerable populations in Northern Nigeria.
   - **Cold-Chain Impact:** Outbreak responses often require immediate mass vaccination campaigns in hot, dry, and rural settings where maintaining a rigid cold chain (2–8°C) is operationally difficult. Engineering the thermal stability of the fHbp antigen is actively studied to support "Controlled Temperature Chain" (CTC) compliance, allowing the vaccine to be stored at ambient temperatures for short periods before administration.
   - **Size:** ~253 to 266 amino acids (mature lipoprotein form).
   - **Structure:** An all-beta folded structure consisting of two beta-barrel domains.
   - **PDB Reference:** 3KVD (wild-type fHbp structure) or 5T5F.
   - **Safety:** Using recombinant fHbp proteins or computational modeling avoids handling the live pathogen, allowing work in BSL-1.

4. **Rotavirus VP8* Core Domain**
   - **Health Challenge:** Rotavirus is a major cause of severe, dehydrating diarrhea and mortality in infants and young children across West Africa.
   - **Cold-Chain Impact:** Current live-attenuated rotavirus vaccines require strict cold chains and are highly sensitive to thermal degradation. Subunit vaccines displaying the receptor-binding VP8* domain are being developed as thermostable alternatives. Engineering the core VP8* domain to resist heat denaturation would pave the way for a highly stable, cost-effective vaccine formulation.
   - **Size:** ~160 amino acids (core galectin-like domain, residues 64–224) or 247 amino acids (full cleaved VP8*).
   - **Structure:** A galectin-like beta-sandwich structure.
   - **PDB Reference:** 2O35 (VP8* core domain) or 1KQR.
   - **Safety:** Work involves a non-infectious capsid subunit protein, which is safe for BSL-1.

**Recommended Next Steps for Your Project**
- For an initial computational study focusing on simple, fast folding predictions and molecular dynamics simulations:
  - Choose **Single-Chain Insulin (SCI)** if your primary goal is to study monomeric stability, unfolding pathways, or preventing aggregation (fibrillation) in a very small, clean model.
  - Choose **PfMSP1-19** if you want to focus on a local infectious disease target and tackle the computational challenge of modeling stability in the presence of multiple disulfide bonds.

</details>

### 1.2 Implementation Prompt

**Prompt:**
> What are the best programmatic resources and Python libraries for fetching these protein sequences and metadata directly into a Google Colab environment? Build a function that achieves this for the Insulin protein you recommended above.


In [ ]:
# @title Code for fetching protein sequences
import requests
import json

# Target UniProt ID (e.g., Human Insulin)
uniprot_id = "P01308"

# 1. Fetch metadata in JSON format
json_url = f"https://rest.uniprot.org/uniprotkb/{uniprot_id}.json"
response = requests.get(json_url)

if response.status_code == 200:
    data = response.json()
    protein_name = data['proteinDescription']['recommendedName']['fullName']['value']
    gene_name = data['genes'][0]['geneName']['value']
    sequence = data['sequence']['value']
    length = data['sequence']['length']
    print(f"Protein: {protein_name}")
    print(f"Gene: {gene_name}")
    print(f"Length: {length} amino acids")
    print(f"Sequence: {sequence}\n")
else:
    print(f"Failed to fetch JSON metadata. Status code: {response.status_code}")

# 2. Fetch raw FASTA sequence
fasta_url = f"https://rest.uniprot.org/uniprotkb/{uniprot_id}.fasta"
fasta_response = requests.get(fasta_url)
if fasta_response.status_code == 200:
    print("FASTA Format:")
    print(fasta_response.text)


Protein: Insulin
Gene: INS
Length: 110 amino acids
Sequence: MALWMRLLPLLALLALWGPDPAAAFVNQHLCGSHLVEALYLVCGERGFFYTPKTRREAEDLQVGQVELGGGPGAGSLQPLALEGSLQKRGIVEQCCTSICSLYQLENYCN

FASTA Format:
>sp|P01308|INS_HUMAN Insulin OS=Homo sapiens OX=9606 GN=INS PE=1 SV=1
MALWMRLLPLLALLALWGPDPAAAFVNQHLCGSHLVEALYLVCGERGFFYTPKTRREAED
LQVGQVELGGGPGAGSLQPLALEGSLQKRGIVEQCCTSICSLYQLENYCN



### 1.3 Implementation Prompt

**Prompt:**
> How can I visualise a protein structure programmatically within this notebook? Please again use the Insulin protein as an example


In [ ]:
!pip install py3Dmol -q

In [ ]:
# @title Code for visualising proteins
import py3Dmol

# 1. Initialize the viewer using the PDB ID for Human Insulin (1MSO)
viewer = py3Dmol.view(query="pdb:1MSO")

# 2. Style the protein chains
# Chain A and Chain B will be colored differently to see the dimeric structure
viewer.setStyle({'chain': 'A'}, {'cartoon': {'color': '#4A90E2'}})  # Blue for Chain A
viewer.setStyle({'chain': 'B'}, {'cartoon': {'color': '#50E3C2'}})  # Green for Chain B

# 3. Highlight Disulfide Bonds (Cysteine residues)
# Since insulin's stability relies heavily on its 3 disulfide bonds,
# we show these Cysteine residues (CYS) as sticks colored by element.
viewer.addStyle({'resn': 'CYS'}, {'stick': {'colorspec': 'yellowCarbon', 'radius': 0.15}})

# 4. Highlight the sulfur-sulfur bonds specifically (optional but helpful)
viewer.addStyle({'elem': 'S'}, {'sphere': {'color': 'yellow', 'radius': 0.4}})

# 5. Set background and camera view
viewer.setBackgroundColor('white')
viewer.zoomTo()  # Automatically adjusts the camera to fit the protein

# 6. Render the interactive widget
viewer.show()


3Dmol.js failed to load for some reason. Please check your browser console for error messages.

## Phase 2: Exploring Protein Structures and Thermostability

*Focus: Identifying engineering targets, predicting novel structures, and visualising mutations.*


### 🔍 2.1 Discovery Prompt

**Prompt:**
> How does the sequence determine the thermostability of a protein? Please give me a scientifically correct description of the different mechanisms involved while avoiding heavy jargon where possible


<details>
<summary>🤖 Show Example Output</summary>

At its core, a protein’s sequence determines its thermostability by dictates how the 3D folded structure balances two opposing forces: the energy holding the protein together (enthalpy) and the natural tendency of molecules to become disorganized (entropy).

At high temperatures, thermal energy makes the atoms vibrate violently. If these vibrations overcome the forces holding the structure together, the protein unfolds (denatures).

Here are the key sequence-dependent mechanisms that nature uses to make proteins heat-resistant:

1. **Tight Packing of the "Water-Hating" (Hydrophobic) Core**
   - **The Concept:** Most proteins have a water-avoiding core. Amino acids with hydrophobic side chains (like Leucine, Isoleucine, and Valine) cluster together inside the protein, away from water.
   - **Mechanism:** Thermostable proteins have sequences optimized to pack these hydrophobic amino acids together with almost no gaps or cavities. This maximizes weak but numerous physical forces (van der Waals interactions) that act like velcro, holding the core tightly together.

2. **Surface Charge Networks (Salt Bridges)**
   - **The Concept:** Opposites attract. Positively charged amino acids (like Arginine and Lysine) attract negatively charged ones (like Glutamic acid and Aspartic acid).
   - **Mechanism:** In heat-resistant proteins, these charged amino acids are frequently arranged on the outer surface of the protein in cooperative networks (called salt bridges). Because water molecules behave differently at high temperatures, these surface electrostatic networks actually become stronger or remain highly stable, holding the outer shell of the protein together.

3. **Covalent Staples (Disulfide Bonds)**
   - **The Concept:** Covalent bonds are the strongest chemical bonds in a protein.
   - **Mechanism:** The sequence can place Cysteine amino acids in positions where they form covalent links (disulfide bridges) with each other. These act like physical staples. When the protein is heated, these staples physically restrict the chain from unraveling, drastically reducing the structural paths it can take to unfold.

4. **Rigidifying Flexible Loops**
   - **The Concept:** Proteins are not rigid blocks; they have flexible "loop" regions connecting structured helices and sheets. Thermal unfolding almost always starts at these flexible loops.
   - **Mechanism:**
     - **Proline (The Rigidifier):** Proline is an amino acid with a rigid, ring-like structure that naturally restricts the movement of the protein backbone. Replacing loop amino acids with Proline limits how much the loop can move, preventing unfolding.
     - **Glycine (The Flexer):** Glycine has no side chain and is highly flexible. Heat-resistant proteins often have fewer Glycine residues in their loop regions to prevent them from wild whipping motions at high temperatures.

5. **Enhanced Hydrogen Bonding**
   - **The Concept:** Hydrogen bonds are weak chemical interactions that form between polar atoms.
   - **Mechanism:** Thermostable proteins typically have a higher density of hydrogen bonds throughout their structure. By carefully positioning polar amino acids (like Serine, Threonine, or Asparagine), the sequence ensures that if one hydrogen bond breaks due to heat, many others are nearby to immediately hold the structure in place.

**Summary**

If you are designing a sequence to be more thermostable, you are generally trying to:
- Fill internal cavities with hydrophobic residues.
- Add charge partnerships (salt bridges) on the surface.
- Staple the structure with disulfide bonds.
- Replace flexible loop residues (less Glycine, more Proline) to restrict movement.

</details>


**Prompt:**
> Can you summarise why starting by mutating surface residues makes sense as an initial direction for a beginner?

<details>
<summary>🤖 Show Example Output</summary>

Here is a quick summary of why targeting surface residues is the perfect starting point for a beginner in protein engineering:

- **High Forgiveness, Low Risk**: The inside core of a protein is extremely fragile; one wrong mutation there will usually cause the whole protein to collapse or misfold. The surface is highly forgiving, allowing beginners to experiment with mutations without instantly destroying the protein's overall 3D shape.
- **Simple, Intuitive Mechanisms**: The most common surface engineering trick is adding "Molecular Velcro" (salt bridges). It is conceptually very easy for a beginner to understand: you just look for two amino acids near each other on the surface and mutate them into opposite electrical charges so they snap together.
- **Easy to Calculate**: Calculating surface exposure is computationally cheap and easy to visualize. Modeling the precise thermodynamic packing of the internal core requires much heavier, more advanced physics simulations.

It serves as the perfect "sandbox" to learn the computational tools before moving on to the high-risk, high-reward challenge of engineering the protein's core!

</details>

### 2.2 Implementation Prompt

**Prompt:**
> Write a function to estimate which residues of our protein are surface-exposed. Then, update our earlier py3Dmol visualisation code to highlight these surface residues in a contrasting colour so I can easily identify potential mutation targets.


In [ ]:
!pip install biopython -q


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.2/3.2 MB 47.3 MB/s eta 0:00:00


In [ ]:
# @title Code for visualising surface-exposed protein regions

import os
import urllib.request
import py3Dmol
from Bio.PDB import PDBParser
from Bio.PDB.SASA import ShrakeRupley

# Theoretical maximum SASA (Ų) for the 20 standard amino acids (Tien et al. 2013)
TIEN_MAX_ASA = {
    'ALA': 129.0, 'ARG': 274.0, 'ASN': 195.0, 'ASP': 193.0, 'CYS': 167.0,
    'GLU': 223.0, 'GLN': 225.0, 'GLY': 104.0, 'HIS': 224.0, 'ILE': 197.0,
    'LEU': 201.0, 'LYS': 236.0, 'MET': 224.0, 'PHE': 240.0, 'PRO': 159.0,
    'SER': 155.0, 'THR': 172.0, 'TRP': 285.0, 'TYR': 263.0, 'VAL': 174.0
}

def get_surface_exposed_residues(pdb_filepath, rsa_threshold=0.20):
    """Calculates the Relative Solvent Accessibility (RSA) for all residues in a PDB file.
    Identifies residues with RSA higher than rsa_threshold as surface-exposed.
    """
    parser = PDBParser(QUIET=True)
    structure = parser.get_structure("protein", pdb_filepath)
    # Compute absolute SASA at the residue level
    sr = ShrakeRupley(probe_radius=1.4, n_points=100)
    sr.compute(structure, level="R")
    surface_residues = []
    core_residues = []
    for model in structure:
        for chain in model:
            chain_id = chain.id
            for residue in chain:
                # Filter out water, ligands (HETATM) and non-standard residues
                if residue.get_id()[0] == ' ':
                    resname = residue.get_resname()
                    resnum = residue.get_id()[1]
                    if resname in TIEN_MAX_ASA:
                        abs_sasa = residue.sasa
                        max_sasa = TIEN_MAX_ASA[resname]
                        rsa = abs_sasa / max_sasa
                        residue_data = {
                            "chain": chain_id,
                            "resnum": resnum,
                            "resname": resname,
                            "abs_sasa": abs_sasa,
                            "rsa": rsa
                        }
                        if rsa >= rsa_threshold:
                            surface_residues.append(residue_data)
                        else:
                            core_residues.append(residue_data)
    return surface_residues, core_residues

# --- Download and Analyze Insulin (PDB: 1MSO) ---
pdb_id = "1MSO"
pdb_filename = f"{pdb_id}.pdb"
pdb_url = f"https://files.rcsb.org/download/{pdb_filename}"

# Download PDB file if not already present
if not os.path.exists(pdb_filename):
    print(f"Downloading {pdb_filename}...")
    urllib.request.urlretrieve(pdb_url, pdb_filename)

# Run the analysis (Threshold set to 20% RSA)
surface_exposed, core = get_surface_exposed_residues(pdb_filename, rsa_threshold=0.20)
print(f"\nAnalysis complete:")
print(f"Total surface-exposed residues (RSA >= 20%): {len(surface_exposed)}")
print(f"Total core residues (RSA < 20%): {len(core)}")

# --- Visualize using py3Dmol ---
with open(pdb_filename, 'r') as f:
    pdb_data = f.read()

viewer = py3Dmol.view(data=pdb_data, format="pdb")
viewer.setBackgroundColor('white')
# Style all residues to a neutral gray/light-blue default (Core representation)
viewer.setStyle({'cartoon': {'color': '#E5E9F0'}})

# Highlight surface-exposed residues in a contrasting Magenta/Orange color
for res in surface_exposed:
    viewer.setStyle(
        {'chain': res['chain'], 'resi': res['resnum']},
        {'cartoon': {'color': '#FF5733'}} # Contrasting Orange-Red
    )

# Retain the representation of critical disulfide bonds as yellow sticks
viewer.addStyle({'resn': 'CYS'}, {'stick': {'colorspec': 'yellowCarbon', 'radius': 0.12}})
viewer.addStyle({'elem': 'S'}, {'sphere': {'color': 'yellow', 'radius': 0.35}})
viewer.zoomTo()
viewer.show()

# Print a subset of the surface residues as potential mutation targets
print("\nSample Surface-Exposed Residues (Potential Mutation Targets):")
print(f"{'Chain':<6}{'Res Num':<8}{'Res Name':<10}{'RSA (Relative Solvent Accessibility, %)':<8}")
for res in sorted(surface_exposed, key=lambda x: (x['chain'], x['resnum']))[:10]:
    print(f"{res['chain']:<6}{res['resnum']:<8}{res['resname']:<10}{res['rsa']*100:.1f}%")



Analysis complete:
Total surface-exposed residues (RSA >= 20%): 24
Total core residues (RSA < 20%): 78


3Dmol.js failed to load for some reason. Please check your browser console for error messages.


Sample Surface-Exposed Residues (Potential Mutation Targets):
Chain Res Num Res Name  RSA (Relative Solvent Accessibility, %)
A     8       THR       32.9%
A     9       SER       37.1%
A     10      ILE       27.0%
A     14      TYR       36.7%
A     18      ASN       57.8%
B     1       PHE       45.9%
B     2       VAL       30.1%
B     17      LEU       32.4%
B     21      GLU       20.4%
B     29      LYS       28.7%


### 🔍 2.3 🧠 [Optional] Discovery Prompt

**Prompt:**
> If I want to fold and visualise an arbitrary, mutated protein string that isn't in standard databases in my colab, what open-source APIs are available for structural prediction?


In [ ]:
# @title 2.4 🧠 [Optional] Protein folding code
import requests
import py3Dmol

def fold_sequence_with_esmfold(sequence: str) -> str:
    """Sends a sequence to the correct ESMFold API endpoint and returns the PDB text."""
    clean_seq = "".join(sequence.split())
    # CORRECT ENDPOINT (Case-sensitive path)
    url = "https://api.esmatlas.com/foldSequence/v1/pdb/"
    headers = {"Content-Type": "text/plain"}
    print(f"Sending sequence ({len(clean_seq)} AA) to ESMFold API...")
    response = requests.post(url, data=clean_seq, headers=headers)
    if response.status_code == 200:
        print("Structure prediction successful!")
        return response.text
    else:
        raise Exception(f"ESMFold API failed (Status {response.status_code}): {response.text}")

# --- Example: Designing a Custom Single-Chain Insulin (SCI) Analog ---
custom_mutated_insulin = (
    "FVNQHLCGSHLVEALYLVCGERGFFYTPKTKTKTKTKT"  # B Chain
    "GGGPRR"                          # Linker
    "GIVEQCCTSICSLYQLENYCN"           # A Chain
)

# 1. Generate PDB string using the corrected API URL
try:
    pdb_string = fold_sequence_with_esmfold(custom_mutated_insulin)
    # 2. Render the generated structure
    viewer = py3Dmol.view(width=500, height=500)
    viewer.addModel(pdb_string, "pdb")
    viewer.setStyle({'cartoon': {'color': 'spectrum'}})
    # Highlight the custom linker (residues 31 to 36) in bright magenta
    viewer.setStyle({'resi': '31-36'}, {'cartoon': {'color': '#FF00FF'}})
    # Show disulfide bonds as sticks
    viewer.addStyle({'resn': 'CYS'}, {'stick': {'colorspec': 'yellowCarbon', 'radius': 0.12}})
    viewer.addStyle({'elem': 'S'}, {'sphere': {'color': 'yellow', 'radius': 0.35}})
    viewer.zoomTo()
    viewer.show()
except Exception as e:
    print(f"Error: {e}")


Sending sequence (65 AA) to ESMFold API...
Structure prediction successful!


3Dmol.js failed to load for some reason. Please check your browser console for error messages.

### 2.5 Implementation Prompt

**Prompt:**
> Let's mutate several of the highlighted surface residues of our Insulin protein to different, more rigid alternatives (e.g., substituting with Proline). Write code to select mutations, predict the resulting structures and visualise the change in structure


In [ ]:
# @title Code for making mutations and visualising change in structure

import os
import requests
import py3Dmol
import Bio.PDB
from Bio.PDB import PDBParser
from Bio.PDB.SASA import ShrakeRupley

# Theoretical maximum SASA (Tien et al. 2013)
TIEN_MAX_ASA = {
    'ALA': 129.0, 'ARG': 274.0, 'ASN': 195.0, 'ASP': 193.0, 'CYS': 167.0,
    'GLU': 223.0, 'GLN': 225.0, 'GLY': 104.0, 'HIS': 224.0, 'ILE': 197.0,
    'LEU': 201.0, 'LYS': 236.0, 'MET': 224.0, 'PHE': 240.0, 'PRO': 159.0,
    'SER': 155.0, 'THR': 172.0, 'TRP': 285.0, 'TYR': 263.0, 'VAL': 174.0
}

# --- Helper Functions ---
def fold_with_esmfold(sequence: str, name: str) -> str:
    """Folds sequence using the correct public ESMFold API."""
    clean_seq = "".join(sequence.split())
    url = "https://api.esmatlas.com/foldSequence/v1/pdb/"
    headers = {"Content-Type": "text/plain"}
    print(f"Folding {name} ({len(clean_seq)} AA) via ESMFold...")
    response = requests.post(url, data=clean_seq, headers=headers)
    if response.status_code == 200:
        return response.text
    else:
        raise Exception(f"ESMFold failed (Status {response.status_code}): {response.text}")

def get_surface_residues(pdb_filepath, rsa_threshold=0.20):
    """Calculates RSA and returns surface-exposed residues (excluding CYS)."""
    parser = PDBParser(QUIET=True)
    structure = parser.get_structure("protein", pdb_filepath)
    sr = ShrakeRupley(probe_radius=1.4, n_points=100)
    sr.compute(structure, level="R")
    surface_residues = []
    for model in structure:
        for chain in model:
            for residue in chain:
                if residue.get_id()[0] == ' ':
                    resname = residue.get_resname()
                    resnum = residue.get_id()[1]
                    # Exclude Cysteine (CYS) to avoid breaking disulfide bridges
                    if resname in TIEN_MAX_ASA and resname != "CYS":
                        rsa = residue.sasa / TIEN_MAX_ASA[resname]
                        if rsa >= rsa_threshold:
                            surface_residues.append({
                                "resnum": resnum,
                                "resname": resname,
                                "rsa": rsa
                            })
    # Sort by highest exposure
    surface_residues.sort(key=lambda x: x['rsa'], reverse=True)
    return surface_residues

def align_structures(ref_pdb_path, sample_pdb_path, output_pdb_path):
    """Aligns sample structure to reference structure using C-alpha atoms."""
    parser = PDBParser(QUIET=True)
    ref_struct = parser.get_structure("ref", ref_pdb_path)
    sample_struct = parser.get_structure("sample", sample_pdb_path)
    ref_atoms = []
    sample_atoms = []
    for ref_res, sample_res in zip(ref_struct[0].get_residues(), sample_struct[0].get_residues(), strict=True):
        if ref_res.get_id()[0] == ' ' and sample_res.get_id()[0] == ' ':
            if 'CA' in ref_res and 'CA' in sample_res:
                ref_atoms.append(ref_res['CA'])
                sample_atoms.append(sample_res['CA'])
    super_imposer = Bio.PDB.Superimposer()
    super_imposer.set_atoms(ref_atoms, sample_atoms)
    super_imposer.apply(sample_struct[0].get_atoms())
    io = Bio.PDB.PDBIO()
    io.set_structure(sample_struct)
    io.save(output_pdb_path)
    return super_imposer.rms

# --- Execution Pipeline ---
# 1. Wild-type Single-Chain Insulin Sequence
wt_sequence = (
    "FVNQHLCGSHLVEALYLVCGERGFFYTPKT"  # B-chain (1-30)
    "GGGPRR"                          # Linker  (31-36)
    "GIVEQCCTSICSLYQLENYCN"           # A-chain (37-57)
)

# 2. Fold WT and save PDB
wt_pdb_text = fold_with_esmfold(wt_sequence, "Wild-Type")
wt_pdb_file = "insulin_wt.pdb"
with open(wt_pdb_file, "w") as f:
    f.write(wt_pdb_text)

# 3. Analyze surface residues of WT
surface_candidates = get_surface_residues(wt_pdb_file)

# 4. Select top 2 surface residues to mutate to Proline (P)
mutations_to_make = surface_candidates[:2]
print(f"\nSelected mutations based on highest surface exposure (RSA):")
for mut in mutations_to_make:
    print(f"- Position {mut['resnum']}: {mut['resname']} (RSA: {mut['rsa']*100:.1f}%) -> PRO")

# 5. Generate mutant sequence string
seq_list = list(wt_sequence)
for mut in mutations_to_make:
    idx = mut['resnum'] - 1  # 0-based index
    seq_list[idx] = 'P'
mut_sequence = "".join(seq_list)
print(f"Mutated Sequence: {mut_sequence}")

# 6. Fold mutant and save PDB
mut_pdb_text = fold_with_esmfold(mut_sequence, "Mutant")
mut_pdb_file = "insulin_mut_raw.pdb"
with open(mut_pdb_file, "w") as f:
    f.write(mut_pdb_text)

# 7. Align Mutant coordinates to WT coordinates to allow visual comparison
aligned_mut_file = "insulin_mut_aligned.pdb"
rmsd = align_structures(wt_pdb_file, mut_pdb_file, aligned_mut_file)
print(f"\nStructural alignment completed. RMSD: {rmsd:.3f} Å")
if rmsd < 1.5:
    print("Fold is successfully conserved (RMSD is low).")
else:
    print("Significant structural rearrangement detected (RMSD is high).")

# 8. Visualise the Overlay in py3Dmol
with open(wt_pdb_file, "r") as f:
    wt_data = f.read()
with open(aligned_mut_file, "r") as f:
    mut_data = f.read()

viewer = py3Dmol.view(width=600, height=500)
viewer.setBackgroundColor('white')
# Model 0: Wild-type (Colored Green)
viewer.addModel(wt_data, "pdb")
viewer.setStyle({'model': 0}, {'cartoon': {'color': '#A3BE8C', 'opacity': 0.85}})
# Model 1: Aligned Mutant (Colored Salmon Red)
viewer.addModel(mut_data, "pdb")
viewer.setStyle({'model': 1}, {'cartoon': {'color': '#BF616A', 'opacity': 0.85}})

# Highlight the mutation sites as spheres on both models
for mut in mutations_to_make:
    resnum = mut['resnum']
    # WT residue in green spheres
    viewer.addStyle({'model': 0, 'resi': resnum}, {'sphere': {'color': '#3B7A57', 'radius': 0.8}})
    # Mutated residue (Proline) in red spheres
    viewer.addStyle({'model': 1, 'resi': resnum}, {'sphere': {'color': '#8B0000', 'radius': 0.8}})

viewer.zoomTo()
viewer.show()


Folding Wild-Type (57 AA) via ESMFold...

Selected mutations based on highest surface exposure (RSA):
- Position 32: GLY (RSA: 96.0%) -> PRO
- Position 21: GLU (RSA: 84.6%) -> PRO
Mutated Sequence: FVNQHLCGSHLVEALYLVCGPRGFFYTPKTGPGPRRGIVEQCCTSICSLYQLENYCN
Folding Mutant (57 AA) via ESMFold...

Structural alignment completed. RMSD: 0.642 Å
Fold is successfully conserved (RMSD is low).


3Dmol.js failed to load for some reason. Please check your browser console for error messages.

**Prompt:**
> What are the takeaways from this visualisation exercise?


<details>
<summary>🤖 Show Example Output</summary>

The takeaways from this are:

1. **The Colors (The Overlay)**

You are looking at two separate 3D models stacked exactly on top of each other:

- Green Ribbon: The original Wild-Type insulin.
- Red/Salmon Ribbon: The Mutant insulin you generated by inserting rigid Prolines.

By superimposing them, you can instantly see if your mutation destroyed the protein. If the red ribbon looked like a chaotic ball of spaghetti while the green one was neatly folded, you would know your mutation caused a catastrophic misfolding.

2. **The Spheres (The Mutation Sites)**

Look at the far left and the far right of the image. You can see clusters of spheres.

- The Green Spheres represent the original flexible amino acids that the script identified as highly surface-exposed.
- The Red Spheres are your new Proline mutations.

Notice where they are located - they are sitting right on the outer edges of the loose, floppy loops. They are completely separated from the dense internal core of the protein.

3.**The Verdict: A Structural Success!**

**The most critical part of this image is how tightly the Red ribbon hugs the Green ribbon**, especially in those coiled alpha-helices near the bottom right.

- This tells you that the ESMFold neural network predicts your mutations did not break the fold.
- Because you successfully targeted highly exposed surface residues on flexible loops, you were able to insert rigid Prolines without causing a "Proline kink" that would shatter the inner helices.

Conclusion: You have theoretically created a more thermostable variant of Insulin! By replacing floppy surface residues with rigid Prolines, you decreased the flexibility of those outer loops. This would in theory make it harder for heat to shake the protein apart, all while preserving the exact 3D shape needed for the protein to function.

### 🔍 2.6 Discovery Prompt

**Prompt:**
> Is there a quantitative way to measure how these mutations affect the stability of the protein? What are the strengths and weaknesses of the best models for this task, and are there any blind spots I should be cautious of


<details>
<summary>🤖 Show Example Output</summary>

**Measuring Stability Quantitatively: ΔΔG**

Yes, there is a strict thermodynamic metric for this. It is called ΔΔG (Delta Delta G), which represents the change in Gibbs Free Energy.

- ΔG is the energy difference between a protein being neatly folded versus unraveled like spaghetti.
- ΔΔG is the difference in that energy between your Wild-Type and your Mutant.
- The Rule: A negative ΔΔG (e.g., -1.5 kcal/mol) means the mutation is stabilizing. A positive ΔΔG means it is destabilizing.

**The Best Models for Predicting ΔΔG**

There are three main categories of tools you can use to calculate this, each with unique strengths and weaknesses:

1. **Physics-Based Force Fields (e.g., Rosetta cartesian_ddg)**
- Strengths: This is the gold standard. Rosetta physically simulates the atomic forces (electrostatics, van der Waals forces, hydrogen bonds) to calculate how much energy it takes to pack the new mutation into the space. It physically rotates the sidechains to find the lowest energy fit.
- Weaknesses: It is computationally massive and extremely slow. You generally need a high-performance computing cluster, not just a Colab notebook.

2. **Empirical Fast Calculators (e.g., FoldX)**
- Strengths: FoldX uses simplified mathematical shortcuts derived from real physical experiments. It is lightning-fast and can scan hundreds of mutations in minutes.
- Weaknesses: It is incredibly sensitive to the exact starting coordinates of the atoms. If your starting PDB file has even a slight clash (two atoms drawn slightly too close together), FoldX will throw a massive error penalty and give you wildly inaccurate results.

3. **Machine Learning Models (e.g., RaSP, MutateX)**
- Strengths: These are the newest generation. Models like RaSP use 3D convolutional neural networks to look at the pocket around the mutation and instantly predict the Rosetta-level ΔΔG in a fraction of a second.
- Weaknesses: Like all AI, they are limited by their training data. If your protein has a highly unusual fold or unusual metals bound to it, the AI might hallucinate an incorrect stability score.


**⚠️ The Blind Spots (Crucial Warnings)**

If you are entering this field, there are two massive blind spots you must avoid:

- **Blind Spot 1: "The AlphaFold Illusion"**

The most common mistake beginners make is assuming that because AlphaFold or ESMFold predicted a beautiful, high-confidence structure for a mutant, the mutant must be stable. This is false. AlphaFold is essentially a highly advanced pattern-recognition engine for finding the final 3D shape; it is terrible at thermodynamics. You can introduce a mutation that completely destroys the protein's hydrophobic core (making it wildly unstable in real life), and AlphaFold will often still confidently draw it perfectly folded. You must use a dedicated ΔΔG tool (like FoldX or RaSP) to prove stability, not a folding algorithm.

- **Blind Spot 2: Stability vs. Aggregation**

You might design a mutation that gives you a fantastic, highly negative ΔΔG. The protein is technically more stable. However, if your mutation exposed a sticky, hydrophobic patch on the surface, the proteins will clump together into a useless sludge (aggregation) when you try to manufacture them in the lab. A protein must be both thermodynamically stable and highly soluble!

</details>


# Phase 3: Building the Optimisation Loop for Discovery

*Focus: Implementing the oracle model, constructing an automated feedback
loop, and evolving from naive to intelligent protein optimization.*

**Prerequisites:** Phases 1 and 2 complete. You should have:
- The wild-type insulin sequence (`wt_sequence`)
- The surface-exposed mutation target identified by SASA analysis
- An understanding of thermostability mechanisms and ΔΔG

**What we'll build:** A closed-loop system where an AI agent proposes
protein mutations, a computational oracle evaluates them, and the system
iteratively improves the protein — all without a wet lab.


## 3.1 Recap: Our Starting Point

From Phase 2, we have our wild-type human insulin sequence and the
surface-exposed mutation target identified by solvent accessibility
analysis.


In [ ]:
# Wild-type human single-chain insulin (SCI, 57 aa) — from Phase 1 & 2
wt_sequence = (
    "FVNQHLCGSHLVEALYLVCGERGFFYTPKT"  # B-chain (1-30)
    "GGGPRR"                          # Linker  (31-36)
    "GIVEQCCTSICSLYQLENYCN"           # A-chain (37-57)
)

# Most exposed residue from Phase 2 SASA analysis
# (your exact position may differ based on the ESMFold prediction)
target_pos = 35  # 1-indexed
print(f'Wild-type sequence ({len(wt_sequence)} aa)')
print(f'Mutation target: position {target_pos} ({wt_sequence[target_pos - 1]})')


Wild-type sequence (57 aa)
Mutation target: position 35 (R)


---
## 3.2 Choosing an Oracle Model

### 🔍 Discovery Prompt

**Prompt:**
> How can I score protein sequence stability on a Google Colab CPU?



<details>
<summary>🤖 Show Example Output</summary>

Computational protein stability evaluation can be performed using
sequence-only models that predict thermal stability or evolutionary
fitness directly from the amino acid sequence.

1. **ThermoFormer:**
   - **How it works:** A transformer-based model trained to predict
     optimal growth temperature (OGT) in °C directly from protein
     sequence. Higher predicted OGT indicates greater thermal stability.
   - **Trade-offs:** Provides a physically interpretable score (°C),
     but requires cloning the ThermoFormer repository and a specific
     version of `transformers` (<4.40.0).

2. **ESM-2 (Evolutionary Scale Modeling):**
   - **How it works:** A protein language model that scores mutations
     by computing the evolutionary log-likelihood of the sequence.
     Higher average log-likelihood indicates better evolutionary fitness.
   - **Trade-offs:** Captures evolutionary constraints and is fast to
     run, but does not directly predict a physical stability quantity.
     Best used as a secondary fitness constraint.

**Chosen Models:** We use **ThermoFormer** (`GinnM/ThermoFormer`) as
the primary thermal stability predictor and **ESM-2**
(`facebook/esm2_t6_8M_UR50D`, 8M parameters) as a secondary
evolutionary fitness constraint.

> **Student Note on Secondary Constraints:**
> A primary stability oracle (like ThermoFormer) is trained to predict
> a physical stability metric (OGT in °C). However, a protein can be
> highly stable but completely non-functional or biologically unrealistic
> if mutations disrupt native interactions or cause aggregation. ESM-2
> computes evolutionary likelihood ("How likely is this sequence to exist
> in nature?"), acting as an evolutionary guardrail to ensure our designed
> proteins remain biologically plausible.

### Implementation Prompt

**Prompt:**
> Great, let's use ThermoFormer for thermal stability and the small
> ESM-2 model for evolutionary fitness. I'm working in Google Colab
> so I may need to install some packages. Can you write code to load
> both models and create scoring functions I can call on a protein
> sequence? I'd like to test them on this single-chain insulin sequence I
> have stored as a variable called wt_sequence.

</details>



In [ ]:
# Install required libraries and clone ThermoFormer
!git clone https://github.com/ginnm/ThermoFormer.git
!pip install -q "transformers<4.40.0" torch biopython

# Maybe restart the runtime so the downgraded transformers is picked up
# import IPython
# IPython.Application.instance().kernel.do_shutdown(True)

Cloning into 'ThermoFormer'...
remote: Enumerating objects: 22, done.
remote: Counting objects: 100% (22/22), done.
remote: Compressing objects: 100% (18/18), done.
remote: Total 22 (delta 3), reused 13 (delta 0), pack-reused 0 (from 0)
Receiving objects: 100% (22/22), 23.70 KiB | 5.92 MiB/s, done.
Resolving deltas: 100% (3/3), done.
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 4.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.8/8.8 MB 80.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 566.4/566.4 kB 29.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.6/3.6 MB 70.0 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
gradio 6.20.0 requires huggingface-hub<2.0,>=1.2.0, but you have huggingface-hub 0.36.2 which is incompatible.
sentence-transformers 5.6.0 requires transformers<6.0.0,>=4.41

In [ ]:
# @title 1. Model and tokenizer loading
 # After restart: set up paths and import everything
import sys
import os
sys.path.append('ThermoFormer')
os.chdir('ThermoFormer')

import torch
from model.modeling_thermoformer import ThermoFormer
from model.tokenization_thermoformer import ThermoFormerTokenizer
from transformers import AutoTokenizer, AutoModelForMaskedLM

# 1. Setup device
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print(f'Using device: {device}')

# 2. Load ThermoFormer (thermal stability predictor — predicts OGT in °C)
print('Loading ThermoFormer...')
tf_tokenizer = ThermoFormerTokenizer()
tf_model = ThermoFormer.from_pretrained('GinnM/ThermoFormer')
tf_model.eval()
tf_model.to(device)
print('ThermoFormer loaded successfully.')

# 3. Load ESM-2 model (small 8M parameter version — evolutionary fitness)
model_name = 'facebook/esm2_t6_8M_UR50D'
print(f'Loading {model_name}...')
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForMaskedLM.from_pretrained(model_name).to(device)
model.eval()
print('ESM-2 loaded successfully.')

Using device: cpu
Loading ThermoFormer...


/usr/local/lib/python3.12/dist-packages/huggingface_hub/file_download.py:949: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/869 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/2.67G [00:00<?, ?B/s]

ThermoFormer loaded successfully.
Loading facebook/esm2_t6_8M_UR50D...


tokenizer_config.json:   0%|          | 0.00/95.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/93.0 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/125 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/775 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/31.4M [00:00<?, ?B/s]

ESM-2 loaded successfully.


In [ ]:
# @title 2. Define thermal stability scoring function
def get_thermal_score(sequence: str) -> float:
  """Predicts optimal growth temperature (OGT) in °C using ThermoFormer."""
  truncate_length = 2048
  if len(sequence) > truncate_length:
    sequence = sequence[:truncate_length]
  inputs = tf_tokenizer(
      [sequence], return_tensors='pt', padding=True, truncation=True
  )
  input_ids = inputs['input_ids'].to(device)
  attention_mask = inputs['attention_mask'].to(device)
  with torch.no_grad():
    outputs = tf_model(input_ids, attention_mask=attention_mask)
  return outputs.predicted_values.cpu().numpy().tolist()[0]


In [ ]:
# @title 3. Define single-position mutation scoring function (ESM-2)
def score_mutation(sequence: str, pos: int, wild_type: str, mutant: str) -> float:
  """Scores a mutation at `pos` (0-indexed) using the pseudo-log-likelihood ratio."""
  assert 0 <= pos < len(sequence), 'Position out of bounds.'
  assert sequence[pos] == wild_type, f'Expected {wild_type} at index {pos}, but found {sequence[pos]}'

  # Mask the target position
  seq_list = list(sequence)
  seq_list[pos] = tokenizer.mask_token
  masked_sequence = ''.join(seq_list)

  # Tokenize and perform inference
  inputs = tokenizer(masked_sequence, return_tensors='pt').to(device)
  with torch.no_grad():
    outputs = model(**inputs)
    logits = outputs.logits  # [batch_size, seq_len, vocab_size]

  # Map sequence position to token position (handles <cls> start token)
  token_pos = 1 + pos

  # Get token IDs
  wt_id = tokenizer.convert_tokens_to_ids(wild_type)
  mut_id = tokenizer.convert_tokens_to_ids(mutant)

  # Extract log probabilities
  log_probs = torch.log_softmax(logits[0, token_pos], dim=-1)

  # Calculate score
  score = log_probs[mut_id] - log_probs[wt_id]
  return score.item()


In [ ]:
# @title 4. Test both models on the wild-type sequence
print('--- ThermoFormer ---')
wt_thermal = get_thermal_score(wt_sequence)
print(f'Wild-type predicted OGT: {wt_thermal:.2f} °C')

print('\n--- ESM-2 Single-Position Scoring ---')
target_index = target_pos - 1
wt_aa = wt_sequence[target_index]
mut_aa = 'A'  # Test mutation to Alanine

score = score_mutation(wt_sequence, target_index, wt_aa, mut_aa)
print(f'Scoring mutation at position {target_pos}: {wt_aa} -> {mut_aa}')
print(f'ESM-2 Mutational Score: {score:.4f}')
if score > 0:
  print('Result: Mutation is predicted to be more favorable than wild-type.')
else:
  print('Result: Mutation is predicted to be less favorable than wild-type.')


Asking to truncate to max_length but no maximum length is provided and the model has no predefined maximum length. Default to no truncation.


--- ThermoFormer ---
Wild-type predicted OGT: 32.97 °C

--- ESM-2 Single-Position Scoring ---
Scoring mutation at position 35: R -> A
ESM-2 Mutational Score: -0.1358
Result: Mutation is predicted to be less favorable than wild-type.


---
## 3.3 Building a General Sequence Scoring Function

### 🔍 Discovery Prompt

**Prompt:**
> How does ESM-2 evaluate and score a whole sequence?


<details>
<summary>🤖 Show Example Output</summary>

ESM-2 outputs logit scores over the entire vocabulary for every sequence position. To score a whole sequence rather than a single-point mutation delta, we compute the average log-probability of all actual residues in the sequence under the model.

Mathematically:
$$ \text{Fitness Score} = \frac{1}{L} \sum_{i=1}^L \log p(x_i \mid X) $$
where $p(x_i \mid X)$ is the model's predicted probability of the correct amino acid $x_i$ at position $i$ given the unmasked sequence context $X$.
Excluded from the average calculation are special tokens like `<cls>`, `<eos>`, and `<pad>`.

Biologically, a higher average log-probability score indicates that the sequence contains amino acid patterns that are more evolutionary favorable or consistent with the natural proteins ESM-2 was trained on. A mutation that increases this score suggests overall stabilization or evolutionary compatibility.

> **Student Note on Special Tokens:**
> - `<cls>` (classification) is a helper token at the start of a sequence that pools global representation.
> - `<eos>` (end-of-sequence) marks the end.
> - `<pad>` (padding) is used to pad sequences to equal length when batching.
> These helper tokens are introduced by the tokenizer and do not represent real biology. We exclude them from calculations to prevent skewing the average log-probability.

### Implementation Prompt

**Prompt:**
> The single-position score tells me about one mutation at a time.
> Can you write a function that scores an entire protein sequence
> using ESM-2 — something like an overall evolutionary fitness
> score for the whole sequence? I want to compare wild-type vs
> mutant sequences.

</details>



In [ ]:
# @title Sequence fitness score function

def get_sequence_fitness_score(sequence: str, model, tokenizer, device) -> float:
  """Calculates the fitness score of a protein sequence using ESM-2.

  The score is computed as the average log-probability of the actual tokens
  present in the sequence, excluding special tokens (<cls>, <eos>, <pad>, etc.).
  """
  # 1. Tokenize the unmasked sequence
  inputs = tokenizer(sequence, return_tensors='pt')
  inputs = {k: v.to(device) for k, v in inputs.items()}
  input_ids = inputs['input_ids']

  # 2. Perform forward pass
  with torch.no_grad():
    outputs = model(**inputs)
    logits = outputs.logits  # Shape: [1, seq_len + 2, vocab_size]

  # 3. Compute log-probabilities over vocabulary at each position
  log_probs = torch.log_softmax(logits, dim=-1)

  # 4. Retrieve the log-probability of the actual token present at each position
  seq_log_probs = torch.gather(log_probs, dim=-1, index=input_ids.unsqueeze(-1)).squeeze(-1)

  # 5. Exclude special tokens from the average
  special_ids_tensor = torch.tensor(tokenizer.all_special_ids, device=device)
  special_tokens_mask = torch.isin(input_ids, special_ids_tensor)
  non_special_mask = ~special_tokens_mask

  # Extract only non-special tokens log-probabilities
  valid_log_probs = seq_log_probs[non_special_mask]

  # Compute average log-probability across all residues
  average_log_prob = valid_log_probs.mean().item()

  return average_log_prob


---
## 3.4 Testing the Oracle on Phase 2 Mutations

### Implementation Prompt

**Prompt:**
> I identified position 35 (an Arginine) as a surface-exposed residue
> from my SASA analysis. Can you make a few test mutations there
> (e.g. to Proline, Alanine, Valine) and score them with both
> ThermoFormer and ESM-2? Show me a comparison table so I can see
> which mutations help or hurt.


In [ ]:
# @title Computing scores for mutations
# Define mutant sequences at the target position (position 35, 0-indexed index 34)
target_index = target_pos - 1
wt_aa = wt_sequence[target_index]

mutations = {
    'Wild-type': wt_sequence,
    f'{wt_aa}{target_pos}P (Proline)': wt_sequence[:target_index] + 'P' + wt_sequence[target_index+1:],
    f'{wt_aa}{target_pos}A (Alanine)': wt_sequence[:target_index] + 'A' + wt_sequence[target_index+1:],
    f'{wt_aa}{target_pos}V (Valine)': wt_sequence[:target_index] + 'V' + wt_sequence[target_index+1:],
}

# Calculate wild-type baseline scores
wt_thermal_score = get_thermal_score(wt_sequence)
wt_esm_score = get_sequence_fitness_score(wt_sequence, model, tokenizer, device)

# Print comparison table
print(f"{'Sequence / Mutation':<25} | {'Thermal (°C)':<14} | {'Δ Thermal':<12} | {'ESM-2 Score':<12} | {'Δ ESM-2':<12}")
print("-" * 85)
for name, seq in mutations.items():
  thermal = get_thermal_score(seq)
  esm = get_sequence_fitness_score(seq, model, tokenizer, device)
  d_thermal = thermal - wt_thermal_score
  d_esm = esm - wt_esm_score
  if name == 'Wild-type':
    print(f"{name:<25} | {thermal:<14.2f} | {d_thermal:<12.2f} | {esm:<12.6f} | {d_esm:<12.6f}")
  else:
    print(f"{name:<25} | {thermal:<14.2f} | {d_thermal:<+12.2f} | {esm:<12.6f} | {d_esm:<+12.6f}")


Sequence / Mutation       | Thermal (°C)   | Δ Thermal    | ESM-2 Score  | Δ ESM-2     
-------------------------------------------------------------------------------------
Wild-type                 | 32.97          | 0.00         | -0.560551    | 0.000000    
R35P (Proline)            | 32.57          | -0.39        | -0.561041    | -0.000490   
R35A (Alanine)            | 31.72          | -1.25        | -0.564489    | -0.003938   
R35V (Valine)             | 31.56          | -1.41        | -0.563518    | -0.002967   


---
## 3.5 Attempt 1: Random Mutation (Baseline)

### 🔍 Discovery Prompt

**Prompt:**
> What is the simplest way to search for mutations that optimize this score, and what are its limits?


<details>
<summary>🤖 Show Example Output</summary>

The simplest strategy for searching sequence space is a randomized hill-climbing search (specifically, **First-Choice Hill Climbing**). At each step:
1. Propose a single random amino acid substitution at a random position.
2. Score the candidate sequence using our oracle.
3. Accept the mutation only if it strictly improves the fitness score compared to the current sequence.

**Limitations:**
- **Low Efficiency:** Most random mutations are neutral or deleterious (harmful to stability/function). Thus, the search takes many steps to find even a minor improvement.
- **No Biological Focus:** It explores positions completely at random without considering structural context (e.g. active sites, surface vs core).
- **Epistasis:** It only makes single steps and can easily get stuck in local optima. Epistasis means the effect of a mutation at position A depends on position B. If moving to a higher peak requires traversing a fitness valley (where both mutations are needed together), a single-point random search will reject the individual mutations along the way and get stuck.



### Implementation Prompt

**Prompt:**
> Before using Gemini, let's establish a baseline. Can you write a
> simple random mutation approach — pick a random position, swap to
> a random amino acid, and only keep it if the ESM-2 fitness score
> improves? Run it for 5 steps on the single-chain insulin sequence so we
> can see how random search performs.


In [ ]:
# @title Code for making random mutations
import random

def random_mutate(sequence: str) -> tuple[str, int, str, str]:
  """Randomly selects a position and mutates it to a different amino acid.

  Args:
    sequence: The input protein sequence.

  Returns:
    A tuple of (mutated_sequence, position, wild_type_aa, mutant_aa).
    Position is 0-indexed.
  """
  amino_acids = 'ACDEFGHIKLMNPQRSTVWY'
  pos = random.randint(0, len(sequence) - 1)
  wt_aa = sequence[pos]

  # Choose a different AA
  possible_mutants = [aa for aa in amino_acids if aa != wt_aa]
  mutant_aa = random.choice(possible_mutants)

  mutated_sequence = sequence[:pos] + mutant_aa + sequence[pos+1:]
  return mutated_sequence, pos, wt_aa, mutant_aa

# Optimization loop (5 steps)
current_seq = wt_sequence
current_score = get_sequence_fitness_score(current_seq, model, tokenizer, device)

print(f"Initial Sequence: {current_seq}")
print(f"Initial Score: {current_score:.6f}\n")

for step in range(1, 6):
  candidate_seq, pos, wt_aa, mutant_aa = random_mutate(current_seq)
  candidate_score = get_sequence_fitness_score(candidate_seq, model, tokenizer, device)

  accepted = False
  if candidate_score > current_score:
    current_seq = candidate_seq
    current_score = candidate_score
    accepted = True

  mutation_str = f"{wt_aa}{pos + 1}{mutant_aa}"
  print(f"Step {step}:")
  print(f"  Proposed Mutation: {mutation_str}")
  print(f"  Candidate Score:   {candidate_score:.6f}")
  print(f"  Accepted:          {accepted}")
  print(f"  Current Best Score: {current_score:.6f}\n")


Initial Sequence: FVNQHLCGSHLVEALYLVCGERGFFYTPKTGGGPRRGIVEQCCTSICSLYQLENYCN
Initial Score: -0.560551

Step 1:
  Proposed Mutation: T44Y
  Candidate Score:   -0.558408
  Accepted:          True
  Current Best Score: -0.558408

Step 2:
  Proposed Mutation: F1W
  Candidate Score:   -0.586395
  Accepted:          False
  Current Best Score: -0.558408

Step 3:
  Proposed Mutation: F1P
  Candidate Score:   -0.557090
  Accepted:          True
  Current Best Score: -0.557090

Step 4:
  Proposed Mutation: N54L
  Candidate Score:   -0.548846
  Accepted:          True
  Current Best Score: -0.548846

Step 5:
  Proposed Mutation: G31T
  Candidate Score:   -0.554508
  Accepted:          False
  Current Best Score: -0.548846



---
## 3.6 Attempt 2: Gemini-Guided Mutation

### 🔍 Discovery Prompt

**Prompt:**
> Can an LLM like Gemini suggest stabilizing mutations?


<details>
<summary>🤖 Show Example Output</summary>

General Large Language Models (like Gemini) can reason about protein properties by drawing upon their vast text-based training data, which includes literature on biophysical rules, sequence conservation, and structural stability heuristics.

In Google Colab, we can configure the `google-genai` SDK, retrieve a user-provided Gemini API key securely from Colab Secrets, and query the model to suggest stabilizing mutations. By prompting the model with a sequence, its current thermal stability score, and instructions to follow biological rules (e.g. improving hydrophobic core packing, introducing salt bridges, or reducing conformational entropy in helices), we can make structured, rational proposals.

> **Student Note on Structured Outputs:**
> To parse Gemini's suggestions programmatically without fragile regex parsing, we can define a Pydantic schema (using `BaseModel`) representing the variables we want (e.g., the mutation string and reasoning). Passing `response_mime_type="application/json"` and the schema to `response_schema` guarantees that Gemini formats its response to match our Python model exactly, making parsing robust.

</details>



### Implementation Prompt

**Prompt:**
> Now I want to use Gemini to suggest mutations instead of random
> ones. I'm in Google Colab and have my API key stored in Colab
> Secrets. Can you write a function that sends the protein sequence
> and its current thermal stability score to Gemini, asks it to
> suggest a single stabilizing mutation with reasoning, and parses
> the response? Then compare one random mutation vs one Gemini
> mutation on the single-chain insulin sequence.


In [ ]:
# Install required libraries for Google Colab
!pip install -q google-genai

In [ ]:
# @title Code for getting mutation suggestions
import json
import re
from google import genai
from google.genai import types
from google.colab import userdata
from pydantic import BaseModel, Field

# Helper to apply a mutation string (e.g. 'R35A') to a sequence
def apply_mutation(sequence: str, mutation: str) -> str:
  if not mutation:
    return sequence
  try:
    wt_aa = mutation[0]
    pos = int(mutation[1:-1]) - 1  # 0-indexed
    mut_aa = mutation[-1]

    seq_list = list(sequence)
    seq_list[pos] = mut_aa
    return "".join(seq_list)
  except Exception as e:
    print(f"Error applying mutation '{mutation}': {e}")
    return sequence

# Configure Gemini API client
try:
  api_key = userdata.get('GEMINI_API_KEY')
  client = genai.Client(api_key=api_key)
  HAS_GEMINI = True
except Exception as e:
  HAS_GEMINI = False
  print("Warning: GEMINI_API_KEY not found in Colab Secrets. Gemini steps will be skipped.")

class MutationSuggestion(BaseModel):
  mutation: str = Field(
      description="The proposed single-point mutation, e.g., 'R35P' (WT residue, 1-based position, mutant residue)"
  )
  reasoning: str = Field(
      description="A brief explanation of the biophysical rationale (e.g. packing, charge, hydrogen bonds, literature context)"
  )

def ask_gemini_for_mutation(
    sequence: str,
    current_score: float,
    tried_mutations: list[dict] = None,
) -> tuple[str, str]:
  """Prompts Gemini to suggest a single point mutation in the sequence."""
  if not HAS_GEMINI:
    return "", "Gemini API key not configured."

  history_str = ""
  if tried_mutations:
    filtered_history = [item for item in tried_mutations if item.get('step', 0) > 0]
    if filtered_history:
      history_str = "\n  Previously tried mutations:\n"
      for item in filtered_history:
        status = "ACCEPTED" if item.get('accepted') else "REJECTED"
        thermal_val = item.get('thermal') if item.get('thermal') is not None else item.get('candidate_thermal')
        score_info = f"Thermal: {thermal_val:.2f} °C" if thermal_val is not None else "Thermal: N/A"
        if 'composite_score' in item:
          score_info += f", Composite Score: {item.get('composite_score'):.4f}"
        history_str += f"  - {item['mutation']}: {status} ({score_info})\n"
        history_str += f"    Reasoning used: {item.get('reasoning')}\n"
      history_str += "\n  CRITICAL: Do NOT propose any of the previously tried mutations listed above. You must suggest a new single-point mutation that is not in the history.\n"

  prompt = f"""
  You are an expert protein engineer. Propose a single point mutation in the following protein sequence to improve its thermal stability.

  Protein Sequence (Human Single-Chain Insulin):
  {sequence}

  Current Thermal Stability Score (predicted OGT): {current_score:.2f} °C
  {history_str}
  Propose exactly ONE mutation. Use biological reasoning (e.g., sequence conservation, secondary structure propensities, hydrophobic packing, or introducing disulfide bonds/salt bridges).
  """

  try:
    response = client.models.generate_content(
        model='gemini-3.1-flash-lite',
        contents=prompt,
        config=types.GenerateContentConfig(
            response_mime_type="application/json",
            response_schema=MutationSuggestion,
            temperature=1.0,
        ),
    )
    suggestion: MutationSuggestion = response.parsed
    return suggestion.mutation.strip(), suggestion.reasoning.strip()
  except Exception as e:
    return "", f"Error querying Gemini: {e}"

# --- Compare Random vs Gemini-guided mutation ---

# 1. Base WT scores
wt_thermal_score = get_thermal_score(wt_sequence)
wt_esm_score = get_sequence_fitness_score(wt_sequence, model, tokenizer, device)

# 2. Random mutation
rand_mut = random_mutate(wt_sequence)  # returns (mutated_seq, pos, wt_aa, mut_aa)
rand_seq, rand_pos, rand_wt_aa, rand_mut_aa = rand_mut
rand_mut_str = f"{rand_wt_aa}{rand_pos + 1}{rand_mut_aa}"
rand_thermal = get_thermal_score(rand_seq)
rand_esm = get_sequence_fitness_score(rand_seq, model, tokenizer, device)

# 3. Gemini mutation
gemini_mut_str, gemini_reason = "", "Gemini API key not set"
gemini_thermal, gemini_esm = 0.0, 0.0
if HAS_GEMINI:
  print("Querying Gemini for guided mutation...")
  gemini_mut_str, gemini_reason = ask_gemini_for_mutation(wt_sequence, wt_thermal_score)
  if gemini_mut_str:
    gemini_seq = apply_mutation(wt_sequence, gemini_mut_str)
    gemini_thermal = get_thermal_score(gemini_seq)
    gemini_esm = get_sequence_fitness_score(gemini_seq, model, tokenizer, device)

# 4. Print comparison
print("\n" + "=" * 110)
print(f"{'Mutation Type':<20} | {'Mutation':<10} | {'Thermal (°C)':<14} | {'Δ Thermal':<12} | {'ESM-2':<12} | {'Reasoning'}")
print("=" * 110)
print(f"{'Wild-Type':<20} | {'-':<10} | {wt_thermal_score:<14.2f} | {0.0:<12.2f} | {wt_esm_score:<12.6f} | {'Baseline'}")
print(f"{'Random':<20} | {rand_mut_str:<10} | {rand_thermal:<14.2f} | {rand_thermal - wt_thermal_score:<+12.2f} | {rand_esm:<12.6f} | {'Random amino acid substitution'}")
if gemini_mut_str:
  print(f"{'Gemini-Guided':<20} | {gemini_mut_str:<10} | {gemini_thermal:<14.2f} | {gemini_thermal - wt_thermal_score:<+12.2f} | {gemini_esm:<12.6f} | {gemini_reason}")
else:
  print(f"{'Gemini-Guided':<20} | {'N/A':<10} | {'N/A':<14} | {'N/A':<12} | {'N/A':<12} | {gemini_reason}")
print("=" * 110)



Mutation Type        | Mutation   | Thermal (°C)   | Δ Thermal    | ESM-2        | Reasoning
Wild-Type            | -          | 32.97          | 0.00         | -0.560551    | Baseline
Random               | Q51T       | 32.92          | -0.05        | -0.556899    | Random amino acid substitution
Gemini-Guided        | N/A        | N/A            | N/A          | N/A          | Gemini API key not set


---
## 3.7 The Optimisation Loop

### Implementation Prompt

**Prompt:**
> Can you put this into a loop? I want to run multiple rounds where
> Gemini proposes a mutation, I score it with ThermoFormer, and
> only keep it if the thermal stability improves — then feed the
> updated sequence and mutation history back to Gemini for the next
> round so it doesn't repeat failed proposals. Run it for 5 steps
> on the single-chain insulin sequence, and print a summary of the
> best sequence found at the end.

<details>
<summary>🤖 Note on Loop Memory </summary>

> A simple memoryless loop has a major limitation: if Gemini proposes a mutation
> that gets rejected because OGT didn't improve, it has no record of that failure
> and may propose the same rejected mutation again. In a production-grade pipeline,
> we maintain a history of previously tried mutations (both accepted and rejected)
> and append them to the Gemini prompt context to guide it away from failed attempts.
</details>


In [ ]:
# @title Code for an optimisation loop
def optimize_protein(initial_sequence: str, steps: int = 5) -> tuple[str, list[dict]]:
  """Runs a closed-loop optimization using Gemini to propose mutations and ThermoFormer as the oracle."""
  current_sequence = initial_sequence
  current_thermal = get_thermal_score(current_sequence)

  history = []
  print(f"Starting closed-loop optimization...")
  print(f"Initial Sequence: {current_sequence}")
  print(f"Initial Thermal Score: {current_thermal:.2f} °C\n")

  for step in range(1, steps + 1):
    print(f"--- Step {step} ---")

    # 1. Ask Gemini for mutation (passing history memory)
    if HAS_GEMINI:
      mutation_str, reasoning = ask_gemini_for_mutation(current_sequence, current_thermal, history)
    else:
      print("Gemini API not available. Skipping step.")
      break

    if not mutation_str:
      print("No mutation proposed or error occurred. Skipping step.")
      continue

    # 2. Apply proposed mutation
    candidate_sequence = apply_mutation(current_sequence, mutation_str)

    # 3. Score candidate sequence with ThermoFormer
    candidate_thermal = get_thermal_score(candidate_sequence)

    # 4. Accept if thermal score strictly improves
    accepted = False
    if candidate_thermal > current_thermal:
      current_sequence = candidate_sequence
      print(f"Mutation {mutation_str} ACCEPTED (Thermal: {current_thermal:.2f} -> {candidate_thermal:.2f} °C)")
      current_thermal = candidate_thermal
      accepted = True
    else:
      print(f"Mutation {mutation_str} REJECTED (Thermal: {candidate_thermal:.2f} <= current best {current_thermal:.2f} °C)")

    print(f"Reasoning: {reasoning}")
    print(f"Current Best Thermal Score: {current_thermal:.2f} °C\n")

    history.append({
        'step': step,
        'mutation': mutation_str,
        'thermal': candidate_thermal,
        'accepted': accepted,
        'reasoning': reasoning,
        'sequence': current_sequence,
        'thermal_score': current_thermal,
    })

  return current_sequence, history

# Execute the optimization loop for 5 steps starting with wild-type sequence
optimized_sequence, optimization_history = optimize_protein(wt_sequence, steps=5)

# Find and print the best sequence details
accepted_history = [h for h in optimization_history if h['accepted']]
if accepted_history:
  best_entry = accepted_history[-1]
  print("\n" + "="*50)
  print("OPTIMIZATION SUMMARY: BEST SEQUENCE FOUND")
  print("="*50)
  print(f"Step:            {best_entry['step']}")
  print(f"Mutation:        {best_entry['mutation']}")
  print(f"Thermal OGT:     {best_entry['thermal']:.2f} °C")
  print(f"Reasoning:       {best_entry['reasoning']}")
  print("="*50)
  print(f"Sequence:\n{optimized_sequence}")
  print("="*50)
else:
  print("\nNo stabilizing mutations were accepted. The wild-type remains the best sequence.")


Starting closed-loop optimization...
Initial Sequence: FVNQHLCGSHLVEALYLVCGERGFFYTPKTGGGPRRGIVEQCCTSICSLYQLENYCN
Initial Thermal Score: 32.97 °C

--- Step 1 ---
Mutation A13V REJECTED (Thermal: 32.74 <= current best 32.97 °C)
Reasoning: Replacing Alanine with Valine at position 13 in the B-chain helix enhances hydrophobic packing within the insulin core, which is a common stabilization strategy for this region of the protein.
Current Best Thermal Score: 32.97 °C

--- Step 2 ---
Mutation L17A REJECTED (Thermal: 28.53 <= current best 32.97 °C)
Reasoning: Replacing Leucine with Alanine at position 17 in the B-chain is intended to relieve steric strain caused by the bulky side chain in the hydrophobic core, potentially improving the packing geometry and increasing the thermal stability of the insulin monomer.
Current Best Thermal Score: 32.97 °C

--- Step 3 ---
Mutation G8A ACCEPTED (Thermal: 32.97 -> 33.55 °C)
Reasoning: Replacing glycine with alanine at position 8 restricts the conformat

**Prompt:**
> Can you explain what happened above?

<details>
<summary>🤖 Show Example Output</summary>

Here is a breakdown of what happened:

**1. The Setup**

The script uses closed-loop "AI Scientist" logic. It asks Gemini to propose a mutation, applies it, and then acts as an "oracle" to judge the result. However, this time, the oracle is exclusively a model called ThermoFormer, which strictly predicts the Optimal Growth Temperature (OGT / melting point).

**2. The Experiment (Step-by-Step)**

- Initialization: It starts with the same wild-type insulin at 32.97 °C.
- Step 1 (G8A - ACCEPTED): Gemini proposes the exact same first move as before—mutating the floppy Glycine to a rigid Alanine to decrease entropy. The ThermoFormer model confirms this works, raising the predicted stability to 33.55 °C.
- Step 2 (S9A - ACCEPTED): Gemini tries replacing the Serine next to it with another Alanine to further stabilize the helix. The score ticks up slightly to 33.65 °C.
- Steps 3, 4, & 5 (REJECTED): Gemini tries some aggressive moves in the core (introducing bulky Leucines and Isoleucines). The ThermoFormer model catches that these mutations actually disrupt the packing and lower the melting temperature, so the script correctly rejects all of them.

**Warning**

Note a major danger in computational biology: Optimization Blindness. Because this script only measures heat resistance, if Gemini had proposed a mutation in Step 3 that boosted the thermal score to 40.00 °C but completely destroyed the protein's ability to bind to the human insulin receptor (or made it highly immunogenic), this script would have gladly accepted it as a "success."

This is exactly why you might want to track other properties of the protein alongside heat resistance.

---
## 3.8 🧠 [Optional / Advanced] Function Preservation

### 🔍 Discovery Prompt

**Prompt:**
> How do I measure if mutations preserve the protein's original function?


<details>
<summary>🤖 Show Example Output</summary>

To ensure our optimization loop doesn't trade biological function for raw thermostability, we can calculate sequence and semantic similarity metrics relative to the wild-type template:

1. **Edit Distance (Levenshtein Distance):**
   - **Concept:** Calculates the minimum number of single-character edits (insertions, deletions, substitutions) to transform the mutated sequence back to wild-type.
   - **Limitation:** It is a raw sequence metric that penalizes all substitutions equally, failing to capture biological conservative swaps (e.g. L to I vs L to D).

2. **Embedding Similarity:**
   - **Concept:** Extracts the hidden states from the final layer of ESM-2 for the unmasked sequences. By mean-pooling these representations over the sequence (excluding special tokens like `<cls>` and `<eos>`), we get a 1D vector representing the structural/functional context of the protein. The cosine similarity between the wild-type and mutant embeddings measures how closely the mutant aligns with the original biological space.
   - **Significance:** A cosine similarity score close to 1.0 (e.g., > 0.99) indicates that the overall fold and biochemical properties are preserved.

> **Student Note on Embedding Vectors:**
> ESM-2 is trained on millions of natural protein sequences. Inside the model, it converts each amino acid token into a vector capturing its context—its likelihood of being near other specific residues, secondary structure context, and chemical properties. Averaging these token vectors yields a 1,280-dimensional representation of the entire sequence. Cosine similarity measures the angle between these vectors; if they point in almost the same direction, the model's structural understanding of the mutated protein is virtually identical to the wild-type.

</details>



### Implementation Prompt

**Prompt:**
> How can I check whether my optimized sequence is still
> functionally similar to the wild-type? I'd like to measure
> both simple sequence distance and a deeper similarity using
> the ESM-2 embeddings. Can you write code to compare the
> wild-type and optimized single-chain insulin sequences?


In [ ]:
# @title Code for computing similarity between two proteins
import numpy as np

def edit_distance(seq1: str, seq2: str) -> int:
  """Calculates the Levenshtein distance between two sequences using dynamic programming."""
  m, n = len(seq1), len(seq2)
  dp = np.zeros((m + 1, n + 1), dtype=int)

  # Initialize base cases
  for i in range(m + 1):
    dp[i, 0] = i
  for j in range(n + 1):
    dp[0, j] = j

  # Fill DP table
  for i in range(1, m + 1):
    for j in range(1, n + 1):
      cost = 0 if seq1[i - 1] == seq2[j - 1] else 1
      dp[i, j] = min(
          dp[i - 1, j] + 1,      # Deletion
          dp[i, j - 1] + 1,      # Insertion
          dp[i - 1, j - 1] + cost # Substitution
      )
  return int(dp[m, n])

def get_mean_pooled_embedding(sequence: str, model, tokenizer, device) -> torch.Tensor:
  """Extracts the mean-pooled embedding for a single sequence, excluding special tokens."""
  inputs = tokenizer(sequence, return_tensors="pt")
  inputs = {k: v.to(device) for k, v in inputs.items()}

  with torch.no_grad():
    # Pass output_hidden_states=True to get final layer representations
    outputs = model(**inputs, output_hidden_states=True)

  # Outputs contain hidden_states if requested
  last_hidden_state = outputs.hidden_states[-1]  # [1, seq_len + 2, hidden_dim]

  attention_mask = inputs["attention_mask"]
  seq_len = int(attention_mask.sum().item())

  # Extract representations of the actual amino acids (excluding <cls> at index 0 and <eos> at seq_len - 1)
  valid_embeddings = last_hidden_state[0, 1:seq_len-1, :]  # [seq_len, hidden_dim]

  # Mean pooling over the sequence length dimension
  mean_embedding = torch.mean(valid_embeddings, dim=0)  # [hidden_dim]
  return mean_embedding

def embedding_similarity(seq1: str, seq2: str, model, tokenizer, device) -> float:
  """Calculates the cosine similarity between the mean-pooled embeddings of two sequences."""
  emb1 = get_mean_pooled_embedding(seq1, model, tokenizer, device)
  emb2 = get_mean_pooled_embedding(seq2, model, tokenizer, device)

  # Calculate cosine similarity
  cos_sim = torch.nn.functional.cosine_similarity(emb1.unsqueeze(0), emb2.unsqueeze(0))
  return float(cos_sim.item())

# Compare WT and the optimized sequence from the optimization loop
distance_wt_opt = edit_distance(wt_sequence, optimized_sequence)
similarity_wt_opt = embedding_similarity(wt_sequence, optimized_sequence, model, tokenizer, device)

print(f"Comparison: Wild-Type vs Optimized Sequence")
print("-" * 50)
print(f"Edit Distance: {distance_wt_opt}")
print(f"ESM-2 Embedding Cosine Similarity: {similarity_wt_opt:.6f}")


Comparison: Wild-Type vs Optimized Sequence
--------------------------------------------------
Edit Distance: 2
ESM-2 Embedding Cosine Similarity: 0.998898


---
## 3.9 🧠 [Optional / Advanced] Adding Evaluation Metrics


### 🔍 Discovery Prompt

**Prompt:**
> Can you explain what is **hyprophobicity** and **immunogenicity** and why they are important in this context?




<details>
<summary>🤖 Show Example Output</summary>

When using AI to engineer a better, more stable protein (like a therapeutic drug), the algorithm must carefully balance two massive biological forces: **Hydrophobicity** and **Immunogenicity**. Here is what they are and why they dictate the success or failure of our engineered sequence.

#### 1. Hydrophobicity (The Folding Glue)
Hydrophobicity literally means "water-fearing." Because proteins exist in the watery environment of our blood and cells, amino acids that are hydrophobic behave exactly like oil in water—they aggressively repel it.
*   **Why it matters here:** Hydrophobicity is the primary engine of protein stability. To escape the surrounding water, all the "oily" hydrophobic amino acids clump together in the very center of the protein, folding the chain into a dense, solid core.
*   **The Engineering Goal:** If our optimization loop mutates a loose amino acid in the core into a bulkier, highly hydrophobic one (like Leucine or Isoleucine), it packs the protein core even tighter. A tightly packed hydrophobic core requires far more heat energy to melt apart, drastically increasing the protein's thermal stability (OGT).

#### 2. Immunogenicity (The Immune System Alarm)
While the *inside* of the protein handles stability, the *surface* handles the immune system. **Immunogenicity** is the measure of how strongly the human body will "freak out" and attack our newly engineered shape.
*   **The Quality Control Mechanism:** The human immune system constantly chops up floating proteins into tiny strings and inspects them. If the string exactly matches the wild-type human proteins you were born with, the immune system ignores it (it recognizes it as safe "self").
*   **Why it matters here:** If our AI mutates the protein to increase heat resistance, it might accidentally create a 3D shape the body has never seen before. The immune system will flag this mutated shape as an alien pathogen (like a virus) and sound a massive inflammatory alarm.

#### The Context of this Notebook: Why We Must Co-Optimize
If a therapeutic drug is highly immunogenic, the body will produce Anti-Drug Antibodies (ADAs) that instantly hunt down and destroy the drug every time it's injected, rendering it useless. Worse, it could trigger severe allergic reactions or dangerous autoimmune diseases.

This is why the multi-objective optimization loop in this notebook is so critical. We cannot just tell the AI to "make it as heat-resistant as possible," because it might mutate the surface beyond recognition to achieve that. By tracking both metrics simultaneously, the AI acts as a digital biochemist—searching for clever **hydrophobic** mutations that lock the protein together against heat, while verifying via the **immunogenicity** score that the protein still looks perfectly normal and safe to passing immune cells!



### 🔍 Discovery Prompt

**Prompt:**
> How do I track hydrophobicity and immunogenicity in Colab?



<details>
<summary>🤖 Show Example Output</summary>

Beyond neural model scores, tracking raw physicochemical descriptors
and immunogenicity risk is crucial to confirm that mutations do not
inadvertently disrupt the protein's chemical properties or make it
immunogenic.

Key metrics to track include:
- **Hydrophobicity (Eisenberg Consensus Scale):** Measures the average
  thermodynamic preference of the protein's residues for hydrophobic
- **Immunogenicity (MHC-I Presentation Score):** Predicts how strongly
  peptide fragments from the protein bind to MHC class I molecules
  (e.g. HLA-A*02:01). High presentation scores indicate the protein
  may trigger an immune response — undesirable for a therapeutic.

We can calculate immunogenicity using `mhcflurry`, which predicts MHC-I peptide
presentation from sequence.

> **Student Note on Allele Choice:**
> We use HLA-A*02:01 because it is the most common MHC Class I allele
> in human populations globally (present in up to 50% of individuals in some
> demographics), making it a great representative model allele for spotting
> general immunogenicity issues. For actual therapeutics, candidates are
> screened against a much larger panel of common HLA alleles representing
> diverse human populations.
</details>



### Implementation Prompt

**Prompt:**
> Can you write a scoring function for immunogenicity?
> I want to predict MHC-I binding. I may need to install some libraries.
> Test it on my wild-type and optimized sequences.


In [ ]:
# Install mhcflurry
!pip install -q mhcflurry
!mhcflurry-downloads fetch models_class1_presentation


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 157.0/157.0 kB 8.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 191.7/191.7 kB 9.9 MB/s eta 0:00:00
Fetching 1/22 downloads from release 2.2.0
DOWNLOAD NAME                             ALREADY DOWNLOADED?    WILL DOWNLOAD NOW?    URL                  
models_class1_pan                         NO                     NO                    https://github.com/openvax/mhcflurry/releases/download/pre-2.0/models_class1_pan.selected.20200610.tar.bz2 
models_class1_presentation                NO                     YES                   https://github.com/openvax/mhcflurry/releases/download/pre-2.0/models_class1_presentation.20200611.tar.bz2 
models_class1_processing                  NO                     NO                    https://github.com/openvax/mhcflurry/releases/download/pre-2.0/models_class1_processing.selected.20200611.tar.bz2 
models_class1_pan_unselected              NO                     NO                    https://g

In [ ]:
# @title Code for computing an immunogenicity score
from mhcflurry import Class1PresentationPredictor

# Load MHCflurry predictor
mhc_predictor = Class1PresentationPredictor.load()

def get_immunogenicity_score(sequence: str) -> float:
  """Predicts immunogenicity by scoring 9-mer MHC-I presentation (HLA-A*02:01)."""
  if len(sequence) < 9:
    return 0.0
  kmers = [sequence[i:i+9] for i in range(len(sequence) - 8)]
  df = mhc_predictor.predict(peptides=kmers, alleles=['HLA-A*02:01'])
  return df['presentation_score'].max()


In [ ]:
# @title Code for computing all evaluation metrics on the sequences

# Test all evaluation metrics on WT and optimized sequences
print(f"{'Metric':<30} | {'Wild-Type':<14} | {'Optimized':<14} | {'Difference':<12}")
print("-" * 78)

wt_thermal = get_thermal_score(wt_sequence)
opt_thermal = get_thermal_score(optimized_sequence)
print(f"{'Thermal Stability (°C)':<30} | {wt_thermal:<14.2f} | {opt_thermal:<14.2f} | {opt_thermal - wt_thermal:<+12.2f}")

wt_esm = get_sequence_fitness_score(wt_sequence, model, tokenizer, device)
opt_esm = get_sequence_fitness_score(optimized_sequence, model, tokenizer, device)
print(f"{'ESM-2 Fitness':<30} | {wt_esm:<14.6f} | {opt_esm:<14.6f} | {opt_esm - wt_esm:<+12.6f}")

wt_immuno = get_immunogenicity_score(wt_sequence)
opt_immuno = get_immunogenicity_score(optimized_sequence)
print(f"{'Immunogenicity (MHC-I max)':<30} | {wt_immuno:<14.4f} | {opt_immuno:<14.4f} | {opt_immuno - wt_immuno:<+12.4f}")


Metric                         | Wild-Type      | Optimized      | Difference  
------------------------------------------------------------------------------
Thermal Stability (°C)         | 32.97          | 33.65          | +0.69       
ESM-2 Fitness                  | -0.560551      | -0.566787      | -0.006236   
Predicting processing.


100%|██████████| 1/1 [00:00<00:00,  2.93it/s]


Predicting affinities.


100%|██████████| 1/1 [00:00<00:00,  6.99it/s]


Predicting processing.


100%|██████████| 1/1 [00:00<00:00, 10.08it/s]


Predicting affinities.


100%|██████████| 1/1 [00:00<00:00,  8.73it/s]

Immunogenicity (MHC-I max)     | 0.9476         | 0.9476         | +0.0000     


---
## 3.10 🧠 [Optional / Advanced] Multi-Objective Optimisation

### 🔍 Discovery Prompt

**Prompt:**
> How can I combine these 3 metrics for multi-objective optimization?



<details>
<summary>🤖 Show Example Output</summary>

To balance multiple competing parameters, we formulate a **Composite
Score Function** that combines all three metrics into a single scalar.

The composite score uses thermal stability as the primary objective,
adds a weighted ESM-2 evolutionary fitness term, and applies a
strong penalty for high immunogenicity:

$$ \text{Composite}(s) = T(s) + 2.0 \cdot E(s) - P_{\text{immuno}} $$

where $P_{\text{immuno}} = 20.0 \cdot I(s)$ if $I(s) > 0.5$, else $0$.

This formulation:
- Prioritizes thermal stability (direct OGT predicted °C value)
- Rewards evolutionary fitness (higher ESM log-likelihood is better)
- Strongly penalizes high immunogenicity that would limit therapeutic use

> **Student Note on Weight Parameters:**
> These weights are set heuristically to scale components appropriately. Because
> physical ranges differ significantly (OGT ranges between 30 and 80, ESM-2 is a
> small negative float like -2.5), we apply scaling factors so one metric
> doesn't completely overwhelm the others. In professional settings, systematically
> searching weights (Pareto frontiers) helps select ideal balances.



### Implementation Prompt

**Prompt:**
> Can you combine all three metrics into a single composite score,
> using the formula: Composite = thermal + 2.0 * fitness - (20.0 *
> immunogenicity if immunogenicity > 0.5 else 0.0)? Then update
> my optimization loop to use it — Gemini proposes mutations
> (passing the full mutation history so it doesn't repeat past
> proposals) and we accept only if the composite score improves.
> Run it for 5 steps, display the history in a table, and print a
> summary of the best sequence found at the end.


In [ ]:
# @title Code for defining a composite score
import pandas as pd

def calculate_composite_score(
    thermal: float,
    esm: float,
    immunogenicity: float,
) -> float:
  """Calculates the three-term composite score for multi-objective optimization.

  Components:
    1. Thermal stability (primary objective, in °C)
    2. ESM-2 evolutionary fitness (weighted by 2.0)
    3. Immunogenicity penalty (20.0× if presentation score > 0.5)
  """
  # 1. Thermal stability is our primary objective
  score = thermal

  # 2. ESM score (log-likelihood); higher is better
  score += esm * 2.0

  # 3. Penalize high immunogenicity
  if immunogenicity > 0.5:
    score -= (immunogenicity * 20.0)

  return score


In [ ]:
# @title Code for optimizing a protein using multiple objectives

def optimize_protein_multi(
    initial_sequence: str,
    steps: int = 5,
) -> tuple[str, list[dict]]:
  """Runs closed-loop multi-objective optimization using Gemini and a three-term composite score."""
  current_sequence = initial_sequence

  # Calculate baseline metrics
  base_thermal = get_thermal_score(current_sequence)
  base_esm = get_sequence_fitness_score(current_sequence, model, tokenizer, device)
  base_immuno = get_immunogenicity_score(current_sequence)

  current_composite = calculate_composite_score(
      base_thermal, base_esm, base_immuno
  )

  history = []

  print(f"Starting Multi-Objective Optimization...")
  print(f"Initial Sequence: {current_sequence}")
  print(f"Init Metrics -> Thermal: {base_thermal:.2f} | ESM: {base_esm:.4f} | Immuno: {base_immuno:.4f}")
  print(f"Init Composite Score: {current_composite:.4f}\n")

  # Log WT baseline in history
  history.append({
      'step': 0,
      'mutation': 'Wild-type',
      'accepted': True,
      'thermal': base_thermal,
      'esm': base_esm,
      'immunogenicity': base_immuno,
      'composite_score': current_composite,
      'reasoning': 'Baseline WT sequence',
  })

  # Track current metrics for the Gemini prompt
  current_thermal = base_thermal

  for step in range(1, steps + 1):
    print(f"--- Step {step} ---")

    # 1. Ask Gemini for mutation (using thermal score and history memory in the prompt)
    if HAS_GEMINI:
      mutation_str, reasoning = ask_gemini_for_mutation(
          current_sequence, current_thermal, history
      )
    else:
      print("Gemini API not available. Skipping step.")
      break

    if not mutation_str:
      print("No mutation proposed or error occurred. Skipping step.")
      continue

    # 2. Apply proposed mutation
    candidate_sequence = apply_mutation(current_sequence, mutation_str)

    # 3. Compute all three metrics for the candidate
    new_thermal = get_thermal_score(candidate_sequence)
    new_esm = get_sequence_fitness_score(
        candidate_sequence, model, tokenizer, device
    )
    new_immuno = get_immunogenicity_score(candidate_sequence)

    new_composite = calculate_composite_score(
        new_thermal, new_esm, new_immuno
    )

    # 4. Accept if composite score strictly improves
    accepted = False
    if new_composite > current_composite:
      current_sequence = candidate_sequence
      current_composite = new_composite
      current_thermal = new_thermal
      accepted = True
      print(f"Mutation {mutation_str} ACCEPTED (Composite: {current_composite:.4f})")
    else:
      print(f"Mutation {mutation_str} REJECTED (Composite: {new_composite:.4f} <= {current_composite:.4f})")

    print(f"  Metrics -> Thermal: {new_thermal:.2f} | ESM: {new_esm:.4f} | Immuno: {new_immuno:.4f}")
    print(f"  Reason: {reasoning}\n")

    history.append({
        'step': step,
        'mutation': mutation_str,
        'accepted': accepted,
        'thermal': new_thermal,
        'esm': new_esm,
        'immunogenicity': new_immuno,
        'composite_score': new_composite,
        'reasoning': reasoning,
    })

  return current_sequence, history

# Run the multi-objective optimization loop for 5 steps
optimized_seq, opt_history = optimize_protein_multi(wt_sequence, steps=5)

# Convert history to DataFrame for representation
df_history = pd.DataFrame(opt_history)
df_history

# Find and print the best sequence details
accepted_history = df_history[df_history['accepted'] == True]
best_entry = accepted_history.iloc[-1]

print("\n" + "="*60)
print("OPTIMIZATION SUMMARY: BEST SEQUENCE FOUND (MAX COMPOSITE SCORE)")
print("="*60)
print(f"Step:            {best_entry['step']}")
print(f"Mutation:        {best_entry['mutation']}")
print(f"Composite Score: {best_entry['composite_score']:.4f}")
print(f"Thermal OGT:     {best_entry['thermal']:.2f} °C")
print(f"ESM-2 Fitness:   {best_entry['esm']:.6f}")
print(f"Immunogenicity:  {best_entry['immunogenicity']:.4f}")
print(f"Reasoning:       {best_entry['reasoning']}")
print("="*60)
print(f"Sequence:\n{optimized_seq}")
print("="*60)


Predicting processing.


100%|██████████| 1/1 [00:00<00:00,  3.14it/s]


Predicting affinities.


100%|██████████| 1/1 [00:00<00:00,  2.82it/s]


Starting Multi-Objective Optimization...
Initial Sequence: FVNQHLCGSHLVEALYLVCGERGFFYTPKTGGGPRRGIVEQCCTSICSLYQLENYCN
Init Metrics -> Thermal: 32.97 | ESM: -0.5606 | Immuno: 0.9476
Init Composite Score: 12.8944

--- Step 1 ---
Predicting processing.


100%|██████████| 1/1 [00:00<00:00,  5.39it/s]


Predicting affinities.


100%|██████████| 1/1 [00:00<00:00,  4.47it/s]


Mutation A13P REJECTED (Composite: 12.6189 <= 12.8944)
  Metrics -> Thermal: 32.78 | ESM: -0.5561 | Immuno: 0.9526
  Reason: Introducing a proline residue at position 13 in the B-chain helical segment can restrict the conformational entropy of the unfolded state and enhance local rigidity, which is a common strategy to increase the thermal stability of alpha-helices in insulin-like peptides.

--- Step 2 ---
Predicting processing.


100%|██████████| 1/1 [00:00<00:00,  6.47it/s]


Predicting affinities.


100%|██████████| 1/1 [00:00<00:00,  5.30it/s]


Mutation L17A REJECTED (Composite: 8.1514 <= 12.8944)
  Metrics -> Thermal: 28.53 | ESM: -0.5593 | Immuno: 0.9629
  Reason: Replacing Leucine with Alanine at position 17 in the B-chain helical segment can optimize hydrophobic packing within the insulin core, potentially reducing steric strain and improving the stability of the B-helix/A-chain interface.

--- Step 3 ---
Predicting processing.


100%|██████████| 1/1 [00:00<00:00,  5.35it/s]

Predicting affinities.

100%|██████████| 1/1 [00:00<00:00,  4.82it/s]


Mutation F24A REJECTED (Composite: 12.0091 <= 12.8944)
  Metrics -> Thermal: 32.08 | ESM: -0.5595 | Immuno: 0.9476
  Reason: Replacing Phenylalanine with Alanine at position 24 in the B-chain (part of the hydrophobic core) can reduce steric hindrance and packing strain, potentially allowing for a more compact and stable folding of the C-terminal B-chain segment against the A-chain.

--- Step 4 ---
Predicting processing.


100%|██████████| 1/1 [00:00<00:00,  9.30it/s]


Predicting affinities.


100%|██████████| 1/1 [00:00<00:00, 10.74it/s]


Mutation G8A ACCEPTED (Composite: 13.4683)
  Metrics -> Thermal: 33.55 | ESM: -0.5641 | Immuno: 0.9476
  Reason: Replacing a glycine at position 8 in the B-chain helical segment with alanine restricts the backbone conformational freedom, thereby decreasing the entropy of the unfolded state and potentially stabilizing the helical structure through improved side-chain interaction within the helix.

--- Step 5 ---
Predicting processing.


100%|██████████| 1/1 [00:00<00:00,  9.78it/s]


Predicting affinities.


100%|██████████| 1/1 [00:00<00:00, 10.19it/s]

Mutation S9E REJECTED (Composite: 12.3560 <= 13.4683)
  Metrics -> Thermal: 32.45 | ESM: -0.5695 | Immuno: 0.9476
  Reason: Introducing a glutamic acid at position 9 in the B-chain helical segment can facilitate the formation of an N-cap interaction or an electrostatic interaction with nearby basic residues, potentially stabilizing the B-chain helix through improved hydrogen bonding or salt-bridge formation.


OPTIMIZATION SUMMARY: BEST SEQUENCE FOUND (MAX COMPOSITE SCORE)
Step:            4
Mutation:        G8A
Composite Score: 13.4683
Thermal OGT:     33.55 °C
ESM-2 Fitness:   -0.564132
Immunogenicity:  0.9476
Reasoning:       Replacing a glycine at position 8 in the B-chain helical segment with alanine restricts the backbone conformational freedom, thereby decreasing the entropy of the unfolded state and potentially stabilizing the helical structure through improved side-chain interaction within the helix.
Sequence:
FVNQHLCASHLVEALYLVCGERGFFYTPKTGGGPRRGIVEQCCTSICSLYQLENYCN


**Prompt:**
> Can you explain what happened above?


<details>
<summary>🤖 Show Example Output</summary>

Here is a plain-English breakdown of exactly what is happening:

**1. The Goal (Multi-Objective Optimization)**

When designing a drug like insulin for the real world, you can't just maximize one thing (like heat resistance) because you might accidentally ruin something else (like making it highly immunogenic, as we discussed earlier).

This script tests every proposed mutation against three critical "objectives" simultaneously:

- Thermal OGT: Predicted melting temperature (Heat resistance).
- ESM Fitness: Does this sequence still mathematically look like a functional, naturally evolved protein?
- Immunogenicity: Will the human immune system attack it?

The script mathematically combines these three test results into a single Composite Score. If a mutation pushes this overall score higher, it's a success!

**2. The Mechanism (The Closed Loop)**
The script runs a 5-step loop acting as a simulated laboratory:

- Hypothesis: It asks the Gemini AI to act as a biochemist. Gemini looks at the sequence, analyzes the current score, and proposes one specific mutation, providing a scientific justification for why it thinks it will work.
- Experiment: The script applies Gemini's mutation to the sequence and runs the heavy computational models (ESM, Thermal predictors, etc.).
- Peer Review: It calculates the new Composite Score. If the score is higher than the previous best, it ACCEPTS the mutation and uses this new, better sequence as the baseline for the next round. If the score is lower, it REJECTS the mutation and throws it in the trash.

**3. Step-by-Step Breakdown of Your Run**

- Initialization: It starts with a Wild-Type single-chain insulin. The baseline Composite Score is 12.89.
- Step 1 (G8A - ACCEPTED): Gemini suggests replacing a flexible Glycine (G) with a stiffer Alanine (A). Its reasoning is solid: removing the floppy Glycine reduces the "entropy" (as we discussed before), locking the helix in place. The computational models agree! The thermal score jumps from 32.97 to 33.55, and the Composite Score rises to 13.46. The mutation is permanently added to the protein.
- Step 2 (F24A - REJECTED): Gemini tries replacing a bulky Phenylalanine with Alanine. The computational tests show this actually destabilizes the protein (Thermal drops to 32.49). The mutation is rejected and rolled back.
- Step 3 (S9A - ACCEPTED): Gemini tries mutating the very next amino acid, replacing Serine (S) with Alanine (A) to further stiffen the helix. It works! Thermal stability ticks up again to 33.65, and the Composite Score reaches 13.56.
- Steps 4 & 5 (REJECTED): Gemini tries replacing structural Leucines with Alanines. The computational models heavily punish this (in Step 4, the thermal score crashes down to 28.76). Both are rejected.

**The Final Result**

At the very bottom, the script prints out the ultimate winner of the 5-round experiment.

Out of 5 ideas, the AI successfully discovered 2 mutations (G8A and S9A) that work synergistically. The final output sequence (FVNQHLCAAHLVE...) is a brand new, theoretically engineered variant of insulin that the models predict will be significantly more heat-stable, while remaining medically safe and non-immunogenic!


---
## 3.11 Summary

We built a complete optimisation loop in three stages:

| Stage | Proposer | Evaluator | Result |
|---|---|---|---|
| Random | `random_mutate()` | ESM-2 only | Mostly deleterious |
| Gemini-guided | `ask_gemini_for_mutation()` | ThermoFormer | Higher hit rate |
| Multi-objective | Gemini | ThermoFormer + ESM-2 + immunogenicity | Balanced |

**The general pattern** (applicable beyond proteins):
1. **Propose** a candidate change (random, heuristic, or LLM-guided)
2. **Evaluate** it with one or more oracle models
3. **Accept or reject** based on a composite fitness criterion
4. **Iterate** — building on accepted improvements
